api/envoy/extensions/filters/http/decompressor/v3/decompressor.proto

```diff
@@ -0,0 +1,63 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.decompressor.v3;
+
+import "envoy/config/core/v3/base.proto";
+import "envoy/config/core/v3/extension.proto";
+
+import "google/protobuf/any.proto";
+import "google/protobuf/wrappers.proto";
+
+import "udpa/annotations/migrate.proto";
+import "udpa/annotations/status.proto";
+import "udpa/annotations/versioning.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.decompressor.v3";
+option java_outer_classname = "DecompressorProto";
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Decompressor]
+// [#extension: envoy.filters.http.decompressor]
+
+message Decompressor {
+  // Common configuration for filter behavior on both the request and response direction.
+  message CommonDirectionConfig {
+    // Runtime flag that controls whether the filter is enabled for decompression or not. If set to false, the
+    // filter will operate as a pass-through filter. If the message is unspecified, the filter will be enabled.
+    config.core.v3.RuntimeFeatureFlag enabled = 1;
+
+    // If set, and the Content-Length header is present, the Decompressor filter will buffer
+    // data frames until the final data frame and will update the Content Length header with
+    // the inflated size.
+    // The setting configures the maximum uncompressed size the filter will buffer before the connection
+    // manager will stop buffering and return a 413 response.
+    // If the setting is absent, the filter will stream inflated data frames as they are received.
+    google.protobuf.UInt32Value max_buffered_bytes = 2 [(validate.rules).uint32 = {gt: 0}];
+  }
+
+  // Configuration for filter behavior on the request direction.
+  message RequestDirectionConfig {
+    CommonDirectionConfig common_config = 1;
+
+    // If set to true, and response decompression is enabled, the filter modifies the Accept-Encoding
+    // request header by appending the decompressor_library's encoding. Defaults to true.
+    google.protobuf.BoolValue advertise_accept_encoding = 2;
+  }
+
+  // Configuration for filter behavior on the response direction.
+  message ResponseDirectionConfig {
+    CommonDirectionConfig common_config = 1;
+  }
+
+  // A decompressor library to use for decompression. Currently only
+  // :ref:`envoy.compression.gzip.compressor<envoy_api_msg_extensions.compression.gzip.decompressor.v3.Gzip>`
+  // is included in Envoy.
+  config.core.v3.TypedExtensionConfig decompressor_library = 1
```

Comment:
If we configure bufferization separately for requests and responses shouldn't we have separate configs for the decompressor library as well?

Reply:
We _could_ do that. And the stats would work as well.... I am not sure if that is increasing the cognitive load of someone trying to configure more.

Reply:
I think I'll leave as is to future proof in case people want to add more complexity.

Reply:
FWIW it seemed odd to me that we don't allow this to be configured differently in each direction? Can you at least clarify this applies to both the request and response configuration depending on whether they are configured or not?

Reply:
It's trivial to configure them independently, i.e allow different decompressors on each direction. If both you and @rojkov like the idea of allowing that, I can make the change.

Reply:
> LGTM modulo the comment about the compression config. Not sure if you are still planning on fixing that or not.

@mattklein123 do you mean this? Sorry, I was waiting for you and/or @rojkov to vote one way or another. But I guess both of you have already expressed interest in it, so I will go ahead and add.

Reply:
IMHO it's a question of composability vs monolithic design. As a proponent of the former I'd prefer two sequential decompressor filters configured differently for different directions and only consider the latter if performance became an issue.

The same applies to buffering: coupled with the buffer filters we can get the same functionality with less code complexity. If the buffer filter is limited to requests only then it makes sense to extend it (and only it) - people might want to buffer responses in scenarios other than decompression (e.g. compression).

So, for now I vote for a comment on how to couple filters for various use cases instead of complicating the filter. Later we can reconsider.

Reply:
I'm fine either way as long as the current state is well documented.

Reply:
Alright, I had added documentation in the proto, but I have expanded with an example. The current proto would lend itself pretty easily to adding per direction config in the future is someone wants it.

lmk if you think this is clear @mattklein123 @rojkov https://github.com/envoyproxy/envoy/pull/11172/commits/858c279d82aab752ab59d14a5aebd2685e671115